src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs

```diff
@@ -0,0 +1,226 @@
+        Counts counts = self._separated._counts;
+        while (true)
+        {
+            Debug.Assert(counts.WaiterCount != 0);
+            Counts newCounts = counts;
+            if (counts.SignalCount != 0)
+            {
+                newCounts.DecrementSignalCount();
+                newCounts.DecrementWaiterCount();
+            }
+
+            // This waiter has woken up and this needs to be reflected in the count of waiters signaled to wake
+            if (counts.CountOfWaitersSignaledToWake != 0)
+            {
+                newCounts.DecrementCountOfWaitersSignaledToWake();
+            }
+
+            Counts countsBeforeUpdate = self._separated._counts.InterlockedCompareExchange(newCounts, counts);
+            if (countsBeforeUpdate == counts)
+            {
+                if (counts.SignalCount != 0)
+                {
+                    we.OnSuccess(self, we.State);
+                    return;
+                }
+                break;
+            }
+
+            counts = countsBeforeUpdate;
+        }
+        // if we get here, we need to keep waiting because the SignalCount above was 0 after we did
+        // the CompareExchange - someone took the signal before us.
+        // FIXME(ak): why is the timeoutMs the same as before? wouldn't we starve? why does LowLevelLifoSemaphore.WaitForSignal not decrement timeoutMs?
+        self.PrepareAsyncWaitCore (we.TimeoutMs, we);
+        // on success calls InternalAsyncWaitSuccess, on timeout calls InternalAsyncWaitTimeout
+    }
+
+    private void PrepareAsyncWaitCore(int timeout_ms, WaitEntry internalWaitEntry)
+    {
+        GCHandle gchandle = GCHandle.Alloc (internalWaitEntry);
+        unsafe {
+            delegate* unmanaged<IntPtr, IntPtr, void> successCallback = &SuccessCallback;
+            delegate* unmanaged<IntPtr, IntPtr, void> timeoutCallback = &TimeoutCallback;
+            PrepareAsyncWaitInternal (lifo_semaphore, timeout_ms, successCallback, timeoutCallback, GCHandle.ToIntPtr(gchandle));
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.InternalCall)]
+    private static extern unsafe void PrepareAsyncWaitInternal(IntPtr semaphore,
+                                                               int timeoutMs,
+                                                               /*delegate* unmanaged<IntPtr, IntPtr, void> successCallback*/ void* successCallback,
+                                                               /*delegate* unmanaged<IntPtr, IntPtr, void> timeoutCallback*/ void* timeoutCallback,
+                                                               IntPtr userData);
+
+    [UnmanagedCallersOnly]
+    private static void SuccessCallback(IntPtr lifoSemaphore, IntPtr userData)
+    {
+        GCHandle gchandle = GCHandle.FromIntPtr(userData);
+        WaitEntry internalWaitEntry = (WaitEntry)gchandle.Target!;
+        gchandle.Free();
```

Comment:
According to the sample in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle.fromintptr?view=net-8.0 I'm not sure it's correct to free the handle here. I think it might be a double-free, unless the caller isn't freeing their handle?

Reply:
Yea this is different from the example in the docs.  In this case the ownership of the GC handle passes from `PrepareAsyncWaitCore` - which creates the GCHandle, to the underlying native wait structure, to the `SuccessCallback` or `TimeoutCallback`. At that point, the callback is the sole owner of the GC Handle (the underlying native wait entry was already freed).  So the callback is responsible for freeing the GCHandle.

In the example in the docs, the call to EnumWindows never owns the GCHandle - it is always owned by the calling function, which is responsible for freeing it.